File: uvaonlinejudge/00540 - Team Queue.py
from collections import deque

MAX_ELEMENTS = 999999

class TeamQueue:
    def __init__(self, team_n):
        self.teams = [deque() for i in range(team_n)]
        self.queue = deque()
    
    def enqueue(self, team, value):
        if len(self.teams[team]) == 0:
            self.queue.append(team)
        self.teams[team].append(value)

    def dequeue(self):
        team = self.queue[0]
        value = self.teams[team].popleft()
        if len(self.teams[team]) == 0:
            self.queue.popleft()
        return value

def read_team_number():
    return int(input())

def read_team():
    return (int(i) for i in input().split()[1:])

def read_command():
    line = input().split()
    if line[0] == 'STOP':
        return None
    return line[0], (int(line[1]) if line[0] == 'ENQUEUE' else None)

if __name__ == '__main__':
    membership = [None] * MAX_ELEMENTS
    for i, team_n in enumerate(iter(read_team_number, 0), start=1):
        print('Scenario #{}'.format(i))
        team_queue = TeamQueue(team_n)
        for team in range(team_n):
            for element in read_team():
                membership[element] = team
        for command, element in iter(read_command, None):
            if command == 'ENQUEUE':
                team_queue.enqueue(membership[element], element)
            else:
                print(team_queue.dequeue())
        print()

File: uvaonlinejudge/00540 - Team Queue.c
#include <stdio.h>
#include <assert.h>
#define MAX_ELEMENTS 1000000
#define MAX_TEAMS 1000
#define MAX_ELEMENTS_IN_TEAM 1000
#define MAX_QUEUE_SIZE MAX_ELEMENTS_IN_TEAM

struct Queue{
    int array[MAX_QUEUE_SIZE];
    int size;
    int pos;
};

struct TeamQueue{
    struct Queue teams[MAX_TEAMS];
    struct Queue queue;
};

struct TeamQueue team_queue;

int membership[MAX_ELEMENTS];

void queue_enqueue(struct Queue *queue, int value){
    assert(queue->size < MAX_QUEUE_SIZE);
    queue->array[(queue->pos + queue->size) % MAX_QUEUE_SIZE] = value;
    queue->size++;
}

int queue_dequeue(struct Queue *queue){
    int value;
    assert(queue->size > 0);
    value = queue->array[queue->pos];
    queue->pos = (queue->pos + 1) % MAX_QUEUE_SIZE;
    queue->size--;
    return value;
}

int queue_top(struct Queue *queue){
    assert(queue->size > 0);
    return queue->array[queue->pos];
}

void team_queue_init(struct TeamQueue *team_queue){
    int i;
    for(i = 0; i < MAX_TEAMS; i++)
        team_queue->teams[i].pos = team_queue->teams[i].size = 0;
    team_queue->queue.pos = team_queue->queue.size = 0;
}

void team_queue_enqueue(struct TeamQueue *team_queue, int team, int element){
    if(team_queue->teams[team].size == 0){
        queue_enqueue(&team_queue->queue, team);
    }
    queue_enqueue(&team_queue->teams[team], element);
}

int team_queue_dequeue(struct TeamQueue *team_queue){
    int team, element;
    team = queue_top(&team_queue->queue);
    element = queue_dequeue(&team_queue->teams[team]);
    if(team_queue->teams[team].size == 0)
        queue_dequeue(&team_queue->queue);
    return element;
}

void read_elements(int team_n, int membership[]){
    int team, i, element;
    for(team = 0; team < team_n; team++){
        scanf("%d", &i);
        while(i-- > 0){
            scanf("%d", &element);
            membership[element] = team;
        }
    }
}

int read_command(char *command, int *element){
    char buffer[20];
    scanf("%s", buffer);
    *command = buffer[0];
    if(*command == 'S')
        return 0;
    if(*command == 'E')
        scanf("%d", element);
    return 1;
}

int main(){
    char command;
    int team_n, argument, i = 1;
    while(scanf("%d", &team_n) && team_n){
        printf("Scenario #%d\n", i++);
        read_elements(team_n, membership);
        team_queue_init(&team_queue);
        while(read_command(&command, &argument)){
            if(command == 'E'){
                team_queue_enqueue(&team_queue, membership[argument], argument);
            }else{
                printf("%d\n", team_queue_dequeue(&team_queue));
            }
        }
        printf("\n");
    }
    return 0;
}
